go/rpcclientlib/obscuro_client.go

```diff
@@ -19,39 +29,228 @@ const (
+	}
+
+	// todo: remove this non-decryption part when we make viewing key encryption mandatory across all tests
+	// extract result from the data as-is, in case we can't decrypt/process it below
+	decrypted, ok := rawResult.([]byte) // if viewing key was nil we try and extract result from the data as-is
+	if !ok {
+		decStr, ok := rawResult.(string)
+		if ok {
+			decrypted = []byte(decStr)
+		}
+	}
+	if c.viewingKey != nil {
+		// method is sensitive, so we decrypt it before unmarshalling the result
+		decrypted, err = c.decryptResponse(rawResult)
+		if err != nil {
+			return fmt.Errorf("failed to decrypt args for rpc call - %w", err)
+		}
+	}
+
+	// process the decrypted result to get the desired type and set it on the result pointer
+	err = c.prepareResult(decrypted, result)
+	if err != nil {
+		return fmt.Errorf("failed to extract result from RPC response: %w", err)
+	}
+
+	return nil
+}
+
+func (c *RPCClient) encryptArgs(args ...interface{}) ([]byte, error) {
+	if len(args) == 0 {
+		return nil, nil
+	}
+
+	paramsJSON, err := json.Marshal(args)
+	if err != nil {
+		return nil, fmt.Errorf("could not json encode request params: %w", err)
+	}
+
+	if c.viewingKey == nil {
+		// todo: remove this when viewing keys are non-optional for testing purposes
+		// if no viewing key is set then we don't encrypt (we wouldn't be able to decrypt result).
+		return paramsJSON, nil
+	}
+	return c.encryptParamBytes(paramsJSON)
+}
+
+func (c *RPCClient) encryptParamBytes(params []byte) ([]byte, error) {
+	encryptedParams, err := ecies.Encrypt(rand.Reader, c.enclavePublicKey, params, nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("could not encrypt request params with enclave public key: %w", err)
+	}
+	return encryptedParams, nil
+}
+
+func (c *RPCClient) decryptResponse(resultBlob interface{}) ([]byte, error) {
+	encryptedResult := common.Hex2Bytes(resultBlob.(string))
+	if c.viewingKey == nil {
+		// todo: remove this when we no longer support disabling viewing keys for testing
+		return encryptedResult, nil
+	}
+	decryptedResult, err := c.viewingKey.Decrypt(encryptedResult, nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("could not decrypt enclave response with viewing key: %w", err)
 	}
+
+	return decryptedResult, nil
 }
 
-func (c *clientImpl) Call(result interface{}, method string, args ...interface{}) error {
-	return c.rpcClient.Call(&result, method, args...)
+// prepareResult tries to cast/unmarshal data into the result pointer, based on its type
+func (c *RPCClient) prepareResult(data []byte, result interface{}) error {
+	switch result := result.(type) {
+	case *string:
+		*result = string(data)
+		return nil
+
+	default:
+		// for any other type we attempt to json unmarshal it
+		return json.Unmarshal(data, result)
+	}
 }
 
-func (c *clientImpl) Stop() {
+func (c *RPCClient) Stop() {
 	c.rpcClient.Close()
 }
+
+func (c *RPCClient) SetViewingKey(viewingKey *ecies.PrivateKey, viewingPubKeyBytes []byte) {
+	c.viewingKey = viewingKey
```

Comment:
@BedrockSquirrel Can we rename this to `c.viewingPrivKey` or similar for clarity?